// es-abstract/operations/2023.js
'use strict';

module.exports = {
	abs: {
		url: 'https://262.ecma-international.org/14.0/#eqn-abs'
	},
	AddEntriesFromIterable: {
		url: 'https://262.ecma-international.org/14.0/#sec-add-entries-from-iterable'
	},
	AddRestrictedFunctionProperties: {
		url: 'https://262.ecma-international.org/14.0/#sec-addrestrictedfunctionproperties'
	},
	AddToKeptObjects: {
		url: 'https://262.ecma-international.org/14.0/#sec-addtokeptobjects'
	},
	AddWaiter: {
		url: 'https://262.ecma-international.org/14.0/#sec-addwaiter'
	},
	AdvanceStringIndex: {
		url: 'https://262.ecma-international.org/14.0/#sec-advancestringindex'
	},
	'agent-order': {
		url: 'https://262.ecma-international.org/14.0/#sec-agent-order'
	},
	AgentCanSuspend: {
		url: 'https://262.ecma-international.org/14.0/#sec-agentcansuspend'
	},
	AgentSignifier: {
		url: 'https://262.ecma-international.org/14.0/#sec-agentsignifier'
	},
	AllocateArrayBuffer: {
		url: 'https://262.ecma-international.org/14.0/#sec-allocatearraybuffer'
	},
	AllocateSharedArrayBuffer: {
		url: 'https://262.ecma-international.org/14.0/#sec-allocatesharedarraybuffer'
	},
	AllocateTypedArray: {
		url: 'https://262.ecma-international.org/14.0/#sec-allocatetypedarray'
	},
	AllocateTypedArrayBuffer: {
		url: 'https://262.ecma-international.org/14.0/#sec-allocatetypedarraybuffer'
	},
	ApplyStringOrNumericBinaryOperator: {
		url: 'https://262.ecma-international.org/14.0/#sec-applystringornumericbinaryoperator'
	},
	ArrayCreate: {
		url: 'https://262.ecma-international.org/14.0/#sec-arraycreate'
	},
	ArraySetLength: {
		url: 'https://262.ecma-international.org/14.0/#sec-arraysetlength'
	},
	ArraySpeciesCreate: {
		url: 'https://262.ecma-international.org/14.0/#sec-arrayspeciescreate'
	},
	AsyncBlockStart: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncblockstart'
	},
	AsyncFromSyncIteratorContinuation: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncfromsynciteratorcontinuation'
	},
	AsyncFunctionStart: {
		url: 'https://262.ecma-international.org/14.0/#sec-async-functions-abstract-operations-async-function-start'
	},
	AsyncGeneratorAwaitReturn: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorawaitreturn'
	},
	AsyncGeneratorCompleteStep: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorcompletestep'
	},
	AsyncGeneratorDrainQueue: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratordrainqueue'
	},
	AsyncGeneratorEnqueue: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorenqueue'
	},
	AsyncGeneratorResume: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorresume'
	},
	AsyncGeneratorStart: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorstart'
	},
	AsyncGeneratorUnwrapYieldResumption: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorunwrapyieldresumption'
	},
	AsyncGeneratorValidate: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncgeneratorvalidate'
	},
	AsyncGeneratorYield: {
		url: 'https://262.ecma-international.org/14.0/#sec-asyncg
