// cli-ux/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Errors = tslib_1.__importStar(require("@oclif/errors"));
const util = tslib_1.__importStar(require("util"));
const base_1 = require("./action/base");
exports.ActionBase = base_1.ActionBase;
const config_1 = require("./config");
exports.config = config_1.config;
exports.Config = config_1.Config;
const deps_1 = tslib_1.__importDefault(require("./deps"));
const exit_1 = require("./exit");
exports.ExitError = exit_1.ExitError;
const Table = tslib_1.__importStar(require("./styled/table"));
exports.Table = Table;
exports.ux = {
    config: config_1.config,
    warn: Errors.warn,
    error: Errors.error,
    exit: Errors.exit,
    get prompt() {
        return deps_1.default.prompt.prompt;
    },
    /**
     * "press anykey to continue"
     */
    get anykey() {
        return deps_1.default.prompt.anykey;
    },
    get confirm() {
        return deps_1.default.prompt.confirm;
    },
    get action() {
        return config_1.config.action;
    },
    get prideAction() {
        return config_1.config.prideAction;
    },
    styledObject(obj, keys) {
        exports.ux.info(deps_1.default.styledObject(obj, keys));
    },
    get styledHeader() {
        return deps_1.default.styledHeader;
    },
    get styledJSON() {
        return deps_1.default.styledJSON;
    },
    get table() {
        return deps_1.default.table;
    },
    get tree() {
        return deps_1.default.tree;
    },
    get open()
